[CAMEL-13812] Split FAQ into separate module under user-manual

aashnajena opened the pull request:
- Created a separate module for FAQ under docs/user-manual/modules. 
- Added nav.doc for FAQ, and moved all related pages in docs/user-manual/modules/FAQ/pages.
- Replaced links to and from the FAQ module in the entire camel repository. I'm not sure how to run link checker locally.

I hope these changes make more sense now. I have made them in accordance with @djencks's comments on the [JIRA issue](https://issues.apache.org/jira/browse/CAMEL-13812)

djencks commented:
I have often failed to run a complete link-check locally :-)
What I do is:
- in my local camel-website checkout:
    - have an antora-playbook-local.yml where I've modified the source urls for the parts I'm working on to point to my local checkout(s).
    - I've been working on 3 branches of camel: latest, 2.x, and 3.0.x. For each issue I've created branches off each of the official branches.
    - In order to be able to see all 3 branches "at once" I have one 'real' or full camel checkout for my 'master-issue-number' branch and used 'git worktree' to check out my other two branches into separate source trees so I can edit them all at once.
    - Since Antora does not yet support multiple work trees, the urls for all these branches have to point to my "main" checkout.  I use HEAD for the main checkout of master, and the actual branch names for the other branches.  This means that although Antora will see even uncommitted changes in HEAD, I have to (locally) check in the changes to the other two branches.
- I also update the package.json scripts to use my alternate playbook.

Then I can run `yarn run build` or other scripts, or `mvn clean install`.

Here's what the content section of my local playbook looks like, from the last issue I worked on:
```
content:
  sources:
    - url: ./../camel
      branches: HEAD
      start_path: docs/user-manual
# eip
    - url: ./../camel
      branches: HEAD
      start_path: core/camel-core-engine/src/main/docs
# languages
    - url: ./../camel
      branches: HEAD
      start_path: core/camel-core-languages/src/main/docs
    - url: ./../camel
      branches: HEAD
      start_path: core/camel-xml-jaxp/src/main/docs
    - url: ./../camel
      branches: camel-3.0.x-issue-14782-task-14784
      start_path: core/camel-base/src/main/docs
    - url: ./../camel
      branches: camel-3.0.x-issue-14782-task-14784
      start_path: core/camel-jaxp/src/main/docs
    - url: ./../camel
      branches: camel-2.x-issue-14782-task-14791
      start_path: camel-core/src/main/docs
# main components doc
    - url: ./../camel
      branches: HEAD, camel-2.x-issue-14782-task-14791, camel-3.0.x-issue-14782-task-14784
      start_path: docs/components

    - url: git@github.com:apache/camel-k.git
      branches: master
      start_path: docs
    - url: git@github.com:apache/camel-quarkus.git
      branches: master
      start_path: docs
    - url: git@github.com:apache/camel-kafka-connector.git
      branches: master
      start_path: docs
    - url: git@github.com:apache/camel-spring-boot.git
      branches: master
      start_path: docs
```

aashnajena commented:
Hey, thanks! I have tried to check as many links as possible via the website. The links across different modules in user-manual and between different components (like user-manual and components) seem to work fine. I will run a more comprehensive check as you mentioned above, but I am only dealing with the master branch for now, I haven't considered other branches. Is that why the PR build is failing?

djencks commented:
- The xref changes look appropriate to me: I looked at a few but certainly not all.
- The PR builds always failed for me too, and it looks to me as if the failure has nothing to do with your work.
- I would REALLY prefer you rename the module to faq rather than FAQ.  ROOT is special, otherwise the usual convention is to use lowercase or camel-case names for modules.  Having FAQ rather than faq in a url looks very peculiar.

aashnajena commented:
Yes, I noted that from your comment on the JIRA issue. I'll update the name in a bit!

aashnajena commented:
@djencks I hope the changes are fine now. I initially thought that since FAQ is an acronym it should be in uppercase, but I understand your point. Thanks for making me aware!

aashnajena commented:
@zregvart @djencks any verdict on this PR?

djencks commented:
The changes FAQ >> faq all look good to me, although I didn't build the PR myself to check it.

Someone sometime should remove the docs/user-manual/fix-eip-xrefs.sh script I added accidentally :-)

I wonder if the three files linked to in the faq section of the nav that aren't in the faq module should be: this would be better judged by someone more familiar with camel than I am.

We're definitely going to need fixes for the 2.x and 3.0.x branches, and I'm fine with applying this now as long as the other branches are fixed soon.

aashnajena commented:
Hey! I think I'll delete those extra files if they're not needed. I would be happy to make the changes for both the other branches once this is merged - I think we will also get to know if there were any issues with this PR once we merge?

aashnajena commented:
Also, even I checked for those three files. Turn out, those three files are linked from other sections in the menu as well - languages under DSL, support under Community etc, which is _probably_ why they are kept with others of their kind.

aashnajena commented:
To be clear, there are several active branches, but these changes only need to be made on these two - 3.0.x and 2.x - branches?

zregvart commented:
> To be clear, there are several active branches, but these changes only need to be made on these two - 3.0.x and 2.x - branches?

I think that we should make them on `2.x` and `3.2.x` that's going to get released really soon (but won't include this). Backporting to 3.2 should be fairly easy just cherry-picking changes from here (we'll squash them) so it's easier.

That way on the website we build 2.x, 3.2.x (latest release), and master (latest version).

I'm just testing this locally to see for any issues, if I don't find any (other than the expected breaking of 2.x 3.0.x) then we can proceed and merge this.

zregvart commented:
I'm seeing errors like this one:

```
[INFO] --- camel-package-maven-plugin:3.2.0-SNAPSHOT:xref-check (xref-check) @ docs ---
[ERROR] Unresolved xrefs:
[ERROR]   manual:ROOT:nav.adoc (/home/zregvart/tmp/camel/docs/user-manual/modules/faq/nav.adoc) at line 2: xref:can-i-get-commercial-support.adoc[Can I get commercial support?]
```

Reported by the `camel-package-maven-plugin:xref-check` plugin, if I run `node_modules/@antora/cli/bin/antora --generator @antora/xref-validator antora-playbook-local-xref-check.yml` no errors are reported. I'll create an issue to follow-up in the `camel-package-maven-plugin`.

aashnajena commented:
`xref:can-i-get-commercial-support.adoc` link is present in `user-manual/modules/faq/nav.adoc` and linked file is present at `/user-manual/modules/faq/pages/can-i-get-commercial-support.adoc`. Both `nav.adoc` and `can-i-get-commercial-support.adoc` are in the same folder `faq`, so xref link should be fine right? The links worked fine when I checked on the website..

aashnajena commented:
I'm trying to build the website with this playbook 

```
content:
  sources:
    - url: git@github.com:aashnajena/camel.git
      branches: split-faq
      start_path: docs/user-manual
# eip
    - url: git@github.com:aashnajena/camel.git
      branches: split-faq
      start_path: core/camel-core-engine/src/main/docs
# core languages
    - url: git@github.com:aashnajena/camel.git
      branches: split-faq
      start_path: core/camel-core-languages/src/main/docs
    - url: git@github.com:aashnajena/camel.git
      branches: split-faq
      start_path: core/camel-xml-jaxp/src/main/docs
    - url: git@github.com:apache/camel.git
      branches: camel-3.0.x
      start_path: core/camel-base/src/main/docs
    - url: git@github.com:apache/camel.git
      branches: camel-3.0.x
      start_path: core/camel-jaxp/src/main/docs
    - url: git@github.com:aashnajena/camel.git
      branches: split-faq-2.x
      start_path: camel-core/src/main/docs
# main components/dataformats doc
    - url: git@github.com:aashnajena/camel.git
      branches: split-faq, split-faq-2.x, split-faq-3.2.x
      start_path: docs/components
```
But I'm getting some errors saying `Duplicate nav: latest@components:ROOT:nav.adoc`
What could be the possible reason for this error? I'm trying to mimic the original playbook with my branches.

zregvart commented:
I think we're very close to getting this merged and the other two pull requests for 3.2.x and 2.x branches. Here is the playbook I'm using to test with, the difference with the one @aashnajena posted is that I removed 3.0.x bits, as we won't be building those for the website.


[antora-playbook.yml.txt](https://github.com/apache/camel/files/4427850/antora-playbook.yml.txt)

zregvart commented on docs/user-manual/modules/ROOT/pages/contributing.adoc:
```diff
@@ -79,7 +79,7 @@ To edit the documentation:
 * If you want to add more documentation check for .adoc in codebase
 
-For more information see https://camel.apache.org/manual/latest/faq/how-does-the-website-work.html[How does the website work] or https://camel.apache.org/manual/latest/faq/how-do-i-edit-the-website.html[How do I edit the website for more details].
+For more information see xref:faq:how-does-the-website-work.adoc[How does the website work] or xref:faq:how-do-i-edit-the-website.html[How do I edit the website for more details].
```
We have a typo here (and seems we always did have)

```suggestion
For more information see xref:faq:how-does-the-website-work.adoc[How does the website work] or xref:faq:how-do-i-edit-the-website.adoc[How do I edit the website for more details].
```

zregvart commented:
> `xref:can-i-get-commercial-support.adoc` link is present in `user-manual/modules/faq/nav.adoc` and linked file is present at `/user-manual/modules/faq/pages/can-i-get-commercial-support.adoc`. Both `nav.adoc` and `can-i-get-commercial-support.adoc` are in the same folder `faq`, so xref link should be fine right? The links worked fine when I checked on the website..
> 
> Can you please help me understand what's the error here?

I'm not seeing this, so not sure, what is here should work.

aashnajena commented:
> > `xref:can-i-get-commercial-support.adoc` link is present in `user-manual/modules/faq/nav.adoc` and linked file is present at `/user-manual/modules/faq/pages/can-i-get-commercial-support.adoc`. Both `nav.adoc` and `can-i-get-commercial-support.adoc` are in the same folder `faq`, so xref link should be fine right? The links worked fine when I checked on the website..
> > Can you please help me understand what's the error here?
> 
> I'm not seeing this, so not sure, what is here should work.

Yes I figured it out, everything seems to work fine for me now. Added all requested changes in this PR as well as for 3.2.x

zregvart commented:
@aashnajena we need to move this file:

`docs/user-manual/modules/ROOT/examples/core/camel-management/src/test/java/org/apache/camel/management/ManagedCustomProcessorTest.java`

to

`docs/user-manual/modules/faq/examples/core/camel-management/src/test/java/org/apache/camel/management/ManagedCustomProcessorTest.java`

We also need this change in the `docs/gulpfile.js`:

```diff
diff --git a/docs/gulpfile.js b/docs/gulpfile.js
index 95e79e44f1d..7aa5c3c4415 100644
--- a/docs/gulpfile.js
+++ b/docs/gulpfile.js
@@ -239,7 +239,7 @@ const extractExamples = function(file, enc, next) {
 }
 
 function deleteExamples(){
-    return del(['user-manual/modules/ROOT/examples/', 'components/modules/ROOT/examples/']);
+    return del(['user-manual/modules/ROOT/examples/', 'user-manual/modules/faq/examples/', 'components/modules/ROOT/examples/']);
 }
 
 function createUserManualExamples() {
@@ -248,6 +248,12 @@ function createUserManualExamples() {
         .pipe(dest('user-manual/modules/ROOT/examples/'));
 }
 
+function createFAQExamples() {
+    return src('user-manual/modules/faq/**/*.adoc')
+        .pipe(through2.obj(extractExamples))
+        .pipe(dest('user-manual/modules/faq/examples/'));
+}
+
 function createEIPExamples() {
     return src('../core/camel-core-engine/src/main/docs/modules/eips/**/*.adoc')
         .pipe(through2.obj(extractExamples))
@@ -271,7 +277,7 @@ const symlinks = parallel(
     series(deleteComponentImageSymlinks, createComponentImageSymlinks)
 );
 const nav = parallel(createComponentNav, createComponentDataFormatsNav, createComponentLanguagesNav, createEIPNav);
-const examples = series(deleteExamples, createUserManualExamples, createEIPExamples, createUserManualLanguageExamples, createComponentExamples);
+const examples = series(deleteExamples, createUserManualExamples, createFAQExamples, createEIPExamples, createUserManualLanguageExamples, createComponentExamples);
 
 exports.symlinks = symlinks;
 exports.nav = nav;
```

oscerd commented:
@zregvart @djencks can you have a look?

djencks commented:
Looks good to me.  It would be good to squash into one commit.  Lets get this in soon before it needs to be rebased:-)
Nice work!

zregvart commented:
I'm doing one final round of testing with this and the two other PRs. I'll update shortly.

aashnajena commented:
Should I squash the commits now?

aashnajena commented:
Squashed commits into one. I'll do this for the other branches as well.

zregvart commented:
Thanks!
